### PR TITLE
allow a user to create and cache GPS

### PR DESCRIPTION
**Pull request by scamden**
@eriwen

### CODE REVIEW

**Comment by eriwen** [comment]
Thanks! Would it be possible to add a test for `createGlobalGPSInstance`?

**Comment by scamden** [comment]
Sure, I guess just test that a n instance of GPS ends up on the stack trace
object?
Sterling Camden

Sent from RelateIQ

On January 15, 2016 at 7:52 AMEric Wendelin<notifications@github.com> wrote:

Thanks! Would it be possible to add a test for createGlobalGPSInstance?

—
Reply to this email directly or view it on GitHub
<https://github.com/stacktracejs/stacktrace.js/pull/150#issuecomment-171997845>
.

**Comment by eriwen** [comment]
Actually, what might be a more consistent approach would be to allow the user to pass in a `StackTraceGPS` instance, just like is done here: https://github.com/stacktracejs/stacktrace-gps/blob/master/stacktrace-gps.js#L153

Should be easy to write a test. Here's a sample: https://github.com/stacktracejs/stacktrace-gps/blob/master/spec/stacktrace-gps-spec.js#L209

**Comment by scamden** [comment]
The only reason I didn't take that approach is that the user never actually
calls new on the main stack trace module like they do with GPS; it's called
as a factory by the library and I wanted to make sure the gps instance
would be intentionally cached for all usages not just calls to fromError
externally if that makes sense. Happy to try another approach though if you
have an idea there.
Sterling Camden

Sent from RelateIQ

On January 15, 2016 at 7:02 PMEric Wendelin<notifications@github.com> wrote:

Actually, what might be a more consistent approach would be to allow the
user to pass in a StackTraceGPS instance, just like is done here:
https://github.com/stacktracejs/stacktrace-gps/blob/master/stacktrace-gps.js#L153

Should be easy to write a test. Here's a sample:
https://github.com/stacktracejs/stacktrace-gps/blob/master/spec/stacktrace-gps-spec.js#L209

—
Reply to this email directly or view it on GitHub
<https://github.com/stacktracejs/stacktrace.js/pull/150#issuecomment-172149817>
.

File: stacktrace.js [hunk 1]
```diff
@@ -157,6 +158,16 @@
             }
         },
 
+          /**
```
Comment (eriwen) [inline stacktrace.js]: Nitpick: whitespace here

File: stacktrace.js [hunk 2]
```diff
@@ -157,6 +158,16 @@
...
+         * Caches an instance of StackTraceGPS to ensure that source maps can use it's client side cache
+         *
+         * @param opts {Object}
+         */
+        createGlobalGPSInstance: function StackTrace$$(opts) {
```
Comment (eriwen) [inline stacktrace.js]: How do we handle this?

```
StackTrace.createGlobalGPSInstance(opts);
StackTrace.fromError(error, differentOpts);
```

Comment (eriwen) [inline stacktrace.js]: Also nitpick about function name `StackTrace$$` here

**Comment by eriwen** [comment]
You make good points about avoiding the need to keep track of StackTrace and a separate GPS instance. 

If one can just call `this.gps = new StackTraceGPS(opts)` and then have that persisted, why wouldn't we do that by default instead of requiring that a user call `createGlobalGPSInstance()`? I fear that `this.gps` might be lost and we're back to square one, but I could be wrong.

File: stacktrace.js (hunk 2 above)
Comment (scamden) [inline stacktrace.js]: whoops ya lol, i'll fix these nits. the opts thing occurred to me, but from what i can tell the opts passed to fromError don't share properties with the opts for stacktrace gps. i def could have missed something there, but if so we could just have them be two separate sets of opts so the fromError opts would have no effect on stackTraceGPS. i realize that could be a breaking change , but for people using this new function they should know the opts they originally used are now cached so any opts passed to fromError have no effect on the GPS instance. unless you have a better thought there

**Comment by scamden** [comment]
so i thought about allowing `this.gps = new StackTraceGPS(opts)`, but i used the method instead to maintain encapsulatio. Like the user shouldn't need to know the name of the internal field we're going to use. that said, it's a very simple use case and i don't feel strongly.

**Comment by scamden** [comment]
closing in favor of using the opts.sourceCache

**Pull request closed without merging**